tests/codegen/naked-fn/x86_64-linux.rs

```diff
@@ -0,0 +1,43 @@
+//@ compile-flags: -C no-prepopulate-passes -Copt-level=0
+//@ needs-asm-support
+//@ only-linux
+//@ only-x86_64
```

Comment:
the output is only currently tested on x86_64 linux. How should we test other targets? I think the test runner skips any tests that don't run on the current platform. Maybe CI runs on more targets, but specifically embedded targets appear untestable here, because we do actually need `core::arch::asm` and hence cannot use `#[no_core]`.

Reply:
What exactly do you want to test? If you want to use the entire same test on different architectures you can set the check prefix

```rust
//@ revisions: x86 aarch64

//@[x86] compile-flags: --target x86_64-unknown-linux-gnu
//@[x86] filecheck-flags: --check-prefix CHECK-X86

//@[aarch64] compile-flags: --target aarch64-unknown-linux-gnu
//@[aarch64] filecheck-flags: --check-prefix CHECK-AARCH64
```

Then unspecified like `CHECK-LABEL` get tested on all platforms, but `CHECK-X86` `CHECK-X86-NEXT` only get tested on one architecture.

Reply:
I think different files are needed, because the assembly is also different. 

What I'd like to check here is that for all architectures (so elf, macos, coff and then thumb too because it has some special logic) the right directives are emitted.

The problem I have is that I don't know what to put in those files if I can't run the tests locally.

Reply:
One file is fine. If you do compiler-flags with `--target` then it should run those all regardless of host, whereas `only-x86_64` or `only-aarch64` only run if the host is that platform. So I think you could do:

- Add revisions (as mentioned above) for all platforms you want to check on, one for each target
- Make one function that only has a simple `nop` or `ret`, something that works on multiple platforms
- For that one function use all of `CHECK-X86_64` `CHECK-AARCH64` `CHECK-X86-MACOS` etc to verify the LLVM directives. A plain `CHECK` should work on all platforms
- Have a `#[cfg(target_arch = "x86_64")]` function that has a more complete example and `CHECK-X86_64` that
- If there are any platforms you believe have specialized behavior just add a function, `cfg` it for that platform, and only use `CHECK-ARCH.

Reply:
LLVM has some better examples of what this looks like than we do, e.g. https://github.com/llvm/llvm-project/blob/639560353408f82ca101e12cf4edafa9a492116b/llvm/test/CodeGen/ARM/half.ll

Reply:
I guess you can also use the revision directly rather than setting `--check-prefix` yourself. See e.g. https://github.com/rust-lang/rust/blob/73a228116ae8c8ce73e309eee8c730ce90feac78/tests/codegen/cast-target-abi.rs

Reply:
cool, that is very powerful. So can I actually run these locally? it complains

```
//@ revisions: linux windows macos thumb
//
//@[linux] compile-flags: --target x86_64-unknown-linux-gnu
//@[linux] filecheck-flags: --check-prefix CHECK-LINUX
//
//@[macos] compile-flags: --target aarch64-apple-darwin
//@[macos] filecheck-flags: --check-prefix CHECK-MACOS
//
//@[windows] compile-flags: --target x86_64-pc-windows-gnu
//@[windows] filecheck-flags: --check-prefix CHECK-WINDOWS
//
//@[thumb] compile-flags: --target thumbv7em-none-eabi
//@[thumb] filecheck-flags: --check-prefix CHECK-THUMB
```

```
error[E0463]: can't find crate for `std`
  |
  = note: the `aarch64-apple-darwin` target may not be installed
  = help: consider adding the standard library to the sysroot with `x build library --target aarch64-apple-darwin`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0463`.
------------------------------------------



failures:
    [codegen] tests/codegen/naked-fn/naked-functions.rs#thumb
    [codegen] tests/codegen/naked-fn/naked-functions.rs#windows
    [codegen] tests/codegen/naked-fn/naked-functions.rs#macos
```

Reply:
It's running them for all targets, but it doesn't built `core` for all targets. Most cross-platform assembly tests are `#![no_core]` for that reason, but obviously you can't use `core::asm` that way.

I haven't actually tried this myself but it looks like you should be able to mark `#![no_core]` and then use `#[rustc_builtin_macro]` to redefine `asm!` and test what you want. See https://github.com/rust-lang/rust/blob/73a228116ae8c8ce73e309eee8c730ce90feac78/tests/codegen/asm-sanitize-llvm.rs#L2 as an example

Reply:
thanks that was very helpful! There are now tests for 4 interesting targets (where the logic does something meaningfully different) plus a separate test for the `#[instruction_set]` attribute (that uses a weird old version of arm, so that got too messy in the one file).